pandas/_libs/hashtable_cpp.pyx

```diff
@@ -0,0 +1,70 @@
+import cython
+import numpy as np
+
+cimport numpy as cnp
+from libc.stdint cimport uint32_t
+from libc.string cimport memcpy
+from libcpp.vector cimport vector
+
+from pandas._libs.khash cimport kh_needed_n_buckets
+
+
+cdef extern from "<functional>" namespace "std" nogil:
+    cdef cppclass hash[T]:
+        hash()
+        size_t operator()
+
+cdef extern from "pandas/vendored/klib/cpp/khash.hpp" namespace "klib" nogil:
+    cdef cppclass KHash[T, Hash, Eq=*, khint_t=*]:
+        T *keys
+        KHash()
+        # TODO: validate we don't need deconstructor
+        # ~KHash()
+        void exist(khint_t x)
+        T &at(khint_t x)
+        khint_t get(const T &)
+        # TODO: make this khint_t
+        # int resize(khint_t)
+        int resize(uint32_t)
+        khint_t put(const T &, int *)
+        # void del(khint_t x)
+
+
+# TODO: de-duplicate from hashtable.pyx
+cdef uint32_t SIZE_HINT_LIMIT = (1 << 20) + 7
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+def unique_label_indices(const cnp.npy_intp[:] labels) -> cnp.ndarray:
+    """
+    Indices of the first occurrences of the unique labels
+    *excluding* -1. equivalent to:
+        np.unique(labels, return_index=True)[1]
+    """
+    cdef:
+        int ret = 0
+        Py_ssize_t i, n = len(labels)
+        KHash[cnp.npy_intp, hash[cnp.npy_intp]] *table = (
+            new KHash[cnp.npy_intp, hash[cnp.npy_intp]]()
+        )
+        cnp.ndarray[cnp.npy_intp, ndim=1] arr
+        vector[cnp.npy_intp] idx = vector[cnp.npy_intp]()
+
+    table.resize(min(kh_needed_n_buckets(n), SIZE_HINT_LIMIT))
+
+    with nogil:
+        for i in range(n):
+            table.put(labels[i], &ret)
+            if ret != 0:
+                # TODO: pandas has a custom resize operation but we
+                # rely on C++ stdlib here - how different are they?
+                idx.push_back(i)
+
+    # TODO: must be a cleaner way to do this?
+    # even arr.data = move(idx.data()) would be better but arr.data is readonly
+    arr = np.empty(idx.size(), dtype=np.intp)
+    memcpy(arr.data, idx.const_data(), idx.size() * sizeof(cnp.npy_intp))
```

Comment:
The `data` member in Cython is read-only and I wasn't sure of any ndarray constructor that would properly manage the lifecycle of a raw data buffer. cc @jbrockmendel in case you know of a better way to do this

Reply:
What would be the advantage of using a vector here over just putting things into the ndarray and resizing it?

Reply:
i think @seberg is the person to ask about this

Reply:
@lithomas1 maybe you could do without it, although this was trying to stay a faithful port of the current codebase which creates a custom templated Vector class

std::vector also has the advantage of working out of the box and using RAII; if you were to do this with a raw buffer it takes a few more steps and requires manual memory management, along with diving into ndarray internals